### PR TITLE
[CARBONDATA-2701] Refactor code to store minimal required info in Block and Blocklet Cache

### PR DESCRIPTION
**Pull request by manishgupta88**
Things done as part of this PR
1. Refactored code to keep only minimal information in block and blocklet cache.
2. Introduced segment properties holder at JVM level to hold the segment properties. As it is heavy object, new segment properties object will be created only when schema or cardinality is changed for a table.
This PR depends on PR #2437 

 - [ ] Any interfaces changed?
 No
 - [ ] Any backward compatibility impacted?
 NA
 - [ ] Document update required?
No
 - [ ] Testing done
Yes       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5637/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5638/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6843/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6845/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5636/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6890/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5670/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6927/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5712/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5700/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6928/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5713/

File: core/src/main/java/org/apache/carbondata/core/util/BlockletDataMapUtil.java [hunk 1]
```diff
@@ -321,4 +328,43 @@ private static boolean isSameColumnSchemaList(List<ColumnSchema> indexFileColumn
...
+    dataOutput.writeShort(columnSchemas.size());
+    for (ColumnSchema columnSchema : columnSchemas) {
+      if (columnSchema.getColumnReferenceId() == null) {
+        columnSchema.setColumnReferenceId(columnSchema.getColumnUniqueId());
+      }
+      columnSchema.write(dataOutput);
+    }
+    byte[] byteArray = stream.toByteArray();
+    // Compress with snappy to reduce the size of schema
+    return Snappy.rawCompress(byteArray, byteArray.length);
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/util/BlockletDataMapUtil.java]: Use compressor factory.

File: core/src/main/java/org/apache/carbondata/core/util/BlockletDataMapUtil.java [hunk 2]
```diff
@@ -321,4 +328,43 @@ private static boolean isSameColumnSchemaList(List<ColumnSchema> indexFileColumn
...
+
+  /**
+   * Read column schema from binary
+   *
+   * @param schemaArray
+   * @throws IOException
+   */
+  public static List<ColumnSchema> readColumnSchema(byte[] schemaArray) throws IOException {
+    // uncompress it.
+    schemaArray = Snappy.uncompress(schemaArray);
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/util/BlockletDataMapUtil.java]: Same as abive

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapFactory.java [hunk 3]
```diff
@@ -17,7 +17,11 @@
 package org.apache.carbondata.core.indexstore.blockletindex;
 
 import java.io.IOException;
-import java.util.*;
+import java.util.ArrayList;
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataMapFactory.java]: Remove unnecessary changes

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5718/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6934/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5706/

File: core/src/main/java/org/apache/carbondata/core/indexstore/BlockletDataMapIndexStore.java [hunk 4]
```diff
@@ -184,6 +185,23 @@ public BlockletDataMapIndexWrapper get(TableBlockIndexUniqueIdentifierWrapper id
...
       TableBlockIndexUniqueIdentifierWrapper tableSegmentUniqueIdentifierWrapper) {
+    BlockletDataMapIndexWrapper blockletDataMapIndexWrapper =
+        getIfPresent(tableSegmentUniqueIdentifierWrapper);
+    if (null != blockletDataMapIndexWrapper) {
+      // clear the segmentProperties cache
+      List<BlockDataMap> dataMaps = blockletDataMapIndexWrapper.getDataMaps();
+      if (null != dataMaps) {
+        String segmentId =
+            tableSegmentUniqueIdentifierWrapper.getTableBlockIndexUniqueIdentifier().getSegmentId();
+        for (BlockDataMap dataMap : dataMaps) {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/indexstore/BlockletDataMapIndexStore.java]: this for loop will run ony once , change this to dataMap.get(0)

File: core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java [hunk 5]
```diff
@@ -16,7 +16,9 @@
  */
 package org.apache.carbondata.core.indexstore.blockletindex;
 
-import java.io.*;
+import java.io.IOException;
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockDataMap.java]: Remove this change

**Comment by kumarvishal09** [comment]
LGTM except few minor comments 
@manishgupta88 Please check

**Comment by manishgupta88** [comment]
@kumarvishal09 ..As discussed with you I have handled these comments as part of PR https://github.com/apache/carbondata/pull/2467
